src/index.ts

```diff
@@ -283,37 +280,46 @@ export class Upload extends Pumpify {
   }
 
   private async startUploading() {
-    const reqOpts = {
-      method: 'PUT',
-      url: this.uri!,
-      headers: {
-        'Content-Range': 'bytes ' + this.offset + '-*/' + this.contentLength
-      }
-    };
-
     const bufferStream = this.bufferStream = new PassThrough();
     const offsetStream = this.offsetStream =
-        new PassThrough({transform: this.onChunk.bind(this)});
+        new Transform({transform: this.onChunk.bind(this)});
     const delayStream = new PassThrough();
+    const requestStreamEmbeddedStream = new PassThrough();
 
-    const requestStream = await this.getRequestStream(reqOpts);
-    this.setPipeline(bufferStream, offsetStream, requestStream, delayStream);
-
-    // wait for "complete" from request before letting the stream finish
     delayStream.on('prefinish', () => {
       this.cork();
     });
 
-    requestStream.on('complete', resp => {
-      if (resp.statusCode < 200 || resp.statusCode > 299) {
+    this.on('response', (resp: GaxiosResponse) => {
+      if (resp.data.error) {
+        this.destroy(resp.data.error);
+        return;
+      }
+
+      if (resp.status < 200 || resp.status > 299) {
         this.destroy(new Error('Upload failed'));
         return;
       }
 
-      this.emit('metadata', resp.body);
+      this.emit('metadata', resp.data);
       this.deleteConfig();
       this.uncork();
     });
+
+    this.setPipeline(bufferStream, offsetStream, delayStream);
+
+    this.pipe(requestStreamEmbeddedStream);
+
+    const reqOpts: GaxiosOptions = {
+      method: 'PUT',
+      url: this.uri,
+      headers: {
+        'Content-Range': 'bytes ' + this.offset + '-*/' + this.contentLength
+      },
+      body: requestStreamEmbeddedStream,
```

Comment:
This style of request complicates things. We have a stream that needs to be sent as the `body` of a request. So, normally, we would want to do:

```js
source.pipe(dest)
```

The equivalent with fetch is:

```js
request({ body: source }).then(resp => {
  resp.data.pipe(dest);
})
```

The `gcs-resumable-upload` class extends Pumpify, so we lose some simplicity by having to pipe ourselves into another stream, as opposed to *being* the stream.

Reply:
Could we simplify this by creating a `createRequestStream` method (or something) in gaxios?

Reply:
Issue opened: https://github.com/JustinBeckwith/gaxios/issues/39